In [2]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")
print("Model loaded successfully!")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

Model loaded successfully!


In [3]:
sentence = "My neighbour built a wall on my land without permission"
embedding = model.encode(sentence)

print("Embedding shape:", embedding.shape)
print("First 10 numbers:", embedding[:10])

Embedding shape: (384,)
First 10 numbers: [ 0.08723566  0.13869463 -0.01778679 -0.04681598  0.0056931  -0.08254907
 -0.06706486 -0.03310909 -0.09993812  0.02247922]


In [4]:
from sentence_transformers import util

sentences = [
    "My neighbour built a wall on my land without permission",
    "Someone encroached on my property",
    "I want to order a pizza",
]

embeddings = model.encode(sentences)

similarity_1_2 = util.cos_sim(embeddings[0], embeddings[1])
print("Similarity (land dispute vs encroachment):", similarity_1_2.item())

similarity_1_3 = util.cos_sim(embeddings[0], embeddings[2])
print("Similarity (land dispute vs pizza):", similarity_1_3.item())

Similarity (land dispute vs encroachment): 0.46553510427474976
Similarity (land dispute vs pizza): 0.02630002051591873


In [5]:
law_descriptions = [
    "Someone deceived you to hand over money or property",
    "A cheque you received bounced due to insufficient funds",
    "Someone entered or used your property without permission",
    "Someone physically hurt you",
]

law_embeddings = model.encode(law_descriptions)

query = "my neighbor built a wall on my land"
query_embedding = model.encode(query)

similarities = util.cos_sim(query_embedding, law_embeddings)
print(similarities)

tensor([[0.3048, 0.0942, 0.3378, 0.1769]])


In [6]:
best_match_idx = similarities.argmax().item()
best_score = similarities[0][best_match_idx].item()

print(f"Best match: '{law_descriptions[best_match_idx]}'")
print(f"Score: {best_score:.3f}")

Best match: 'Someone entered or used your property without permission'
Score: 0.338


In [7]:
test_queries = [
    "I lent someone money and they never paid me back",
    "my business partner cheated me out of profits",
    "someone slapped me during a fight",
    "asdkjhaskjdh random gibberish text",
]

for q in test_queries:
    q_emb = model.encode(q)
    sims = util.cos_sim(q_emb, law_embeddings)
    best_idx = sims.argmax().item()
    best_score = sims[0][best_idx].item()
    print(f"\nQuery: {q}")
    print(f"Best match: {law_descriptions[best_idx]} (score: {best_score:.3f})")


Query: I lent someone money and they never paid me back
Best match: Someone deceived you to hand over money or property (score: 0.459)

Query: my business partner cheated me out of profits
Best match: Someone deceived you to hand over money or property (score: 0.310)

Query: someone slapped me during a fight
Best match: Someone physically hurt you (score: 0.601)

Query: asdkjhaskjdh random gibberish text
Best match: Someone deceived you to hand over money or property (score: 0.095)


In [8]:
import time

start = time.time()
_ = model.encode("test query")
end = time.time()
print(f"Single encoding took {end - start:.3f} seconds")

Single encoding took 0.011 seconds


## Sentence-Transformers Exploration — Findings

- Used `all-MiniLM-L6-v2` model — 384-dimension embeddings.
- Confirmed semantic matching works: [paste your actual similarity scores from Step 9 and 10 here]
- Gibberish input scored [your actual number] vs real queries scoring [your actual range] — 
  confirms a similarity threshold can filter out nonsense input.
- Single query encoding took [your actual time] seconds.
- Ready to build the full LEGAL_SECTIONS matching pipeline using this approach in Week 3.